example/gosumtype_1_declaration.boilerplate.go

```diff
@@ -126,3 +199,89 @@ func UserMap[A any](value User, variants UserVariantsMap[A]) A {
 	})
 	return result
 }
+
+// User = Anonymous | Member | Admin
+type User struct {
+	user user
+}
+
+// user is the interface for UserVariants
+type user interface {
+	Match(variants UserVariants)
+}
+
+func (s User) Match(variants UserVariants) {
+	s.user.Match(variants)
+}
+func (s User) MarshalJSON() (data []byte, err error) {
+	s.user.Match(UserVariants{
+		Anonymous: func(arg0Arg PaymentMethod) {
+			data, err = json.Marshal([]any{
+				"Anonymous",
+				anonymousUserVariants{
+					Arg0: arg0Arg,
+				}})
+		},
+		Member: func(emailArg string, sinceArg time.Time) {
+			data, err = json.Marshal([]any{
+				"Member",
+				memberUserVariants{
+					Email: emailArg,
+					Since: sinceArg,
+				}})
```

Comment:
json: each variant is encoded in `["TypeName", { ... data... }]` format

note: this arrangement follows the implementation of https://github.com/choonkeat/elm-auto-encoder-decoder